src/greedy/stable_matching.rs

```diff
@@ -0,0 +1,199 @@
+use std::collections::{HashMap, VecDeque};
+
+pub fn stable_matching(
```

Comment:
First of all, I will propose the way we can decompose the `stable_matching` function into helper functions along with their functionality. Based on the logic of the function, we can split it and place these small chunks into these helper functions.

- `initialize_men`: Initializes men’s proposal data.
- `initialize_women`: Initializes women’s partnership statuses.
- `precompute_woman_ranks`: Precomputes rankings of men for each woman.
- `process_proposal`: Processes a man’s proposal to a woman.
- `woman_prefers_new_man`: Determines if a woman prefers a new proposal over her current engagement.
- `engage_man`: Updates the engagement between a man and a woman.
- `finalize_matches`: Collects and returns the final stable matches.